### ISSUE TITLE
LittleTiles rendering with shaders issue

### ISSUE THREAD

**Issue by CreativeMD** [sp614x/optifine#365]
![](http://i.imgur.com/cnSUY5a.png)

I did some testing and ensured that the vertex formats used have a size of 56. So it seem to be caused by something else. I think my VertexConsumer skips the new added padding elements and only fills the buffer with the known elements. I don't know if that is an issue or if those will be filled later on, but keep in mind that i'm adding the data once the chunk cache gets uploaded. 
Maybe Optifine does something after building the chunk cache (https://github.com/CreativeMD/LittleTiles/blob/1.10.2/src/main/java/com/creativemd/littletiles/client/render/LittleChunkDispatcher.java#L109)?

**Comment by sp614x** [comment]
For the shaders there is some post-processing after the vertex data is added - calculation of normal and mid UV coords. These are stored in the extended part of the vertex format.
Take a look at "shadersmod.client.SVertexBuilder", for example "endAddVertex()" and "endAddVertexData()"

**Comment by CreativeMD** [comment]
Is there a way to get the source of Optifine? I'm using jd-gui, but that is rather annoying (due to obfuscated names)

Anyways, seems to be rather easy to solve this problem. I only have to pass the buffer to some of those methods (if i got it right), So which ones do I have to call, could not find a reference where Optifine uses those methods (might be a bug of jd-gui).

**Comment by sp614x** [comment]
The VertexBuffer shaders changes: https://gist.github.com/sp614x/70d3f9d631104217dc7dcf6e7e285786

**Comment by CreativeMD** [comment]
Wait, if you have done those changes to the VertexBuffer already, why doesn't it work?
https://github.com/CreativeMD/LittleTiles/blob/1.10.2/src/main/java/com/creativemd/littletiles/client/render/LittleChunkDispatcher.java#L197
or is it something you are going to implement? Or do you create an extra class extending the VertexBuffer and use that one instead?

**Comment by CreativeMD** [comment]
I have implemented it and nothing has changed. So guess something else is wrong here.

Tested it using Ultra-D4 for 1.10.2 by creating an OptifineVertex buffer which does the exact some thing as your shown code.

**Comment by CreativeMD** [comment]
Hey,
it's been a while and i just wanted to ask if there is anything new or if you have a different idea of what could be the problem.

**Comment by sp614x** [comment]
No idea.

**Comment by CreativeMD** [comment]
Okay, actually i have no idea as well ...
The only option I see is if you would give you parts of your source code, so i could try to find out what could possibly cause this issue. I would need the source code creating the rendering cache and uploading it only. There is something missing and by looking at your source i might find out what.

So if you feel comfortable with it i would be really happy to take a look at parts of your source code.

**Comment by CreativeMD** [comment]
Fixed it ... after hours of testing ...
The missing data was about block properties (id, meta, rendertype). So i could fix it by feeding my VertexBuffer with the following methods before rendering `SVertexBuilder.pushEntity(blockstate, blockpos, world, buffer);` and `SVertexBuilder.popEntity(buffer)` after rendering.

I also had to make sure Optifine does not override the data again once i will add it to the chunk buffer. I implemented a hacky way to prevent Optifine from doing anything in `addVertexData(int[])`.

But the block was still rendered almost black, after a lot of pain i found out that my renderer tries to get the normal vector out of the data, as i removed it everything worked just fine.

Thanks for your help ... i'm just happy it is working now.

![](http://i.imgur.com/2eRFfRX.png)
